internals/overlord/logstate/manager.go

```diff
@@ -100,13 +102,57 @@ func (m *LogManager) ServiceStarted(service *plan.Service, buffer *servicelog.Ri
 	}
 
 	m.buffers[service.Name] = buffer
+	var envMap map[string]string
+
 	for _, gatherer := range m.gatherers {
 		target := m.plan.LogTargets[gatherer.targetName]
 		if !service.LogsTo(target) {
 			continue
 		}
-		gatherer.ServiceStarted(service, buffer)
+
+		if envMap == nil {
```

Comment:
Why is the creation of `envMap` done inside the loop, instead of once at the top? If you're concerned about efficiency you can make `parseEnv` return a nil map if env is nil -- but I don't think micro-efficiency matters here, as this only happens once per service start, so I'd probably just keep it simple. (You're already evaluating the env vars every loop anyway.)